Content.Server/GameObjects/Components/Construction/RefinableComponent.cs

```diff
@@ -0,0 +1,53 @@
+﻿using System.Threading.Tasks;
+using Content.Server.GameObjects.Components.Interactable;
+using Content.Server.GameObjects.Components.Stack;
+using Content.Shared.GameObjects.Components.Interactable;
+using Content.Shared.Interfaces.GameObjects.Components;
+using Robust.Shared.GameObjects;
+using Robust.Shared.Serialization;
+using Robust.Shared.ViewVariables;
+
+namespace Content.Server.GameObjects.Components.Construction
+{
+    /// <summary>
+    /// Used for something that can be refined by welder.
+    /// For example, glass shard can be refined to glass sheet.
+    /// </summary>
+    [RegisterComponent]
+    public class RefinableComponent : Component, IInteractUsing
+    {
+        [ViewVariables]
+        private string[] _refineResult;
+        [ViewVariables]
+        private float _refineTime;
+
+        public override string Name => "Refinable";
+
+        public override void ExposeData(ObjectSerializer serializer)
+        {
+            base.ExposeData(serializer);
+            serializer.DataField(ref _refineResult, "refineResult", new string[] { "GlassStack" });
+            serializer.DataField(ref _refineTime, "refineTime", 2f);
+        }
+
+        public async Task<bool> InteractUsing(InteractUsingEventArgs eventArgs)
+        {
+            // check if object is welder
+            if (!eventArgs.Using.TryGetComponent(out ToolComponent tool)) return false;
+            if (!await tool.UseTool(eventArgs.User, Owner, _refineTime, ToolQuality.Welding)) return false;
```

Comment:
Look at crates for reference (I think it got merged) and block the weld if someone's already welding it.